Add a blog post about the repo.scala-sbt.org incident (#1492)
scala/scala-lang: blog/_posts/2023-04-20-sbt-plugins-community-repository.md

---
layout: blog-detail
post-type: blog
by: Julien Richard-Foy, Scala Center
title: "sbt Plugins Community Repository"
---

On the 7th of April 2023, many Scala builds stopped working because an outage affected the sbt community repository.
In this blog post, we explain what the sbt community repository is and what it stores, report on the incident,

> and outline what we can do to avoid such problems in the future.

Reviewer: The addition is good.

How about this?

> In this blog post, we describe the sbt community repository, report on the incident,
and outline actions taken so far. There will be more action in the future and a followup blog post.

> and outline what we can do to avoid such problems in the future.

Author: Done in 3fbaa6d

## The Incident

On the 7th of April 2023, many projects of the Scala ecosystem could not anymore resolve
sbt plugins from the community repository `repo.scala-sbt.org` (examples
[here](https://github.com/sbt/sbt-dynver/issues/239#issuecomment-1499791434),
[here](https://github.com/playframework/playframework/issues/11675#issuecomment-1499869916),
[here](https://github.com/sbt/sbt/issues/7204), and
[here](https://github.com/spray/sbt-revolver/issues/112)).

## Retrospective

To understand what happened, here is some background information. sbt plugins are packaged
as modules published to Ivy or Maven repositories. Historically, sbt plugins used to be
published primarily to the sbt community repository hosted on Bintray. However, in 2021,

> JFrog has [sunsetted the Bintray

Reviewer: omit "has"

> JFrog has [sunsetted the Bintray

Reviewer: omit "has"

service](https://jfrog.com/blog/into-the-sunset-bintray-jcenter-gocenter-and-chartcenter/).
Since then, JFrog has granted the Scala Center a cloud-hosted Artifactory instance for free.

> As it was explained in the [release notes of sbt

Reviewer: omit "it was"

1.5.1](https://eed3si9n.com/bintray-to-jfrog-artifactory-migration-status-and-sbt-1.5.1),

> this Artifactory instance has been used as a read-only repository to host all the sbt

Reviewer: s/has been/is/

plugins that were published to the community repository at that time, but the recommendation

> to plugin authors was to switch to the Central Maven repository.

Reviewer 2: I'm pretty sure it's called the Maven Central repository, not Central Maven.
```suggestion
to plugin authors was to switch to the Maven Central repository.
```
Same in several places in the post.

Author: I think it’s called just “Central”. And it’s a Maven repository, hence “the Central Maven repository”. It could be abbreviated to “the Central repository”, but I left “Maven” for clarity. See https://central.sonatype.org/

Reviewer: In practice, everybody calls it “Maven Central”. I'm not sure if I've ever heard anyone call it anything else — I don't think I have.

Author: OK, I’ve changed that.

In practice, the Artifactory instance sponsored by JFrog expires automatically after a
year, and we have to explicitly ask for renewal. What happened the [7th of
April](https://github.com/sbt/sbt/issues/7202) is that the instance expired without
prior notice. Thankfully, Eugene Yokota, the maintainer of sbt, promptly asked JFrog to
renew the Artifactory instance, which they generously did. After a couple more technical
steps, everything was back to normal.

## Current State

Thanks to the efforts of Eugene Yokota and the support from the company JFrog, the situation
is now back to normal: the read-only repository `repo.scala-sbt.org` backed by an Artifactory
instance works again and all the Scala builds should be able to resolve their plugin
dependencies as before.

However, as mentioned by Seth Tisue in [a
comment](https://github.com/sbt/sbt/issues/7202#issuecomment-1500220344), the event
highlighted a weakness in the Scala ecosystem infrastructure.

In practice, what do we rely on the sbt community repository for?

> The sbt community repository is primarily used for two purposes:

Reviewer: has two primary purposes

---
layout: blog-detail
post-type: blog
by: Julien Richard-Foy, Scala Center
title: "sbt Plugins Community Repository"
---

> The 7th of April 2023, many Scala builds stopped working because an outage affected the sbt community repository.

Reviewer 3: ```suggestion
On the 7th of April 2023, many Scala builds stopped working because an outage affected the sbt community repository.
```

> In this blog post, we explain what the sbt community repository is and what it stores, we report on the incident
> that happened, and we outline what we can do to avoid such problems in the future.

Reviewer 3: ```suggestion
In this blog post, we explain what the sbt community repository is and what it stores, report on the incident
that happened, and outline what we can do to avoid such problems in the future.
```

Reviewer: you can also omit "that happened", that is inherent to being an incident :-)

## The Incident

> The 7th of April 2023, many projects of the Scala ecosystem could not anymore resolve

Reviewer 3: ```suggestion
On the 7th of April 2023, many projects in the Scala ecosystem could not resolve
```

In practice, the Artifactory instance sponsored by JFrog expires automatically after a
year, and we have to explicitly ask for renewal. What happened the [7th of
April](https://github.com/sbt/sbt/issues/7202) is that the instance expired without
prior notice. Thankfully, Eugene Yokota, the maintainer of sbt, promptly asked JFrog to
renew the Artifactory instance, which they generously did. After a couple more technical
steps, everything was back to normal.

## Current State

Thanks to the efforts of Eugene Yokota and the support from the company JFrog, the situation
is now back to normal: the read-only repository `repo.scala-sbt.org` backed by an Artifactory
instance works again and all the Scala builds should be able to resolve their plugin
dependencies as before.

However, as mentioned by Seth Tisue in [a
comment](https://github.com/sbt/sbt/issues/7202#issuecomment-1500220344), the event
highlighted a weakness in the Scala ecosystem infrastructure, which is well illustrated by
this comic:

<p style="text-align: center">
  <img style="max-width: 385px" src="https://imgs.xkcd.com/comics/dependency.png" alt="Dependency" />

> </p>

Reviewer 4: I don't think this comic really serves value in the context of the incident postmortem, in part because it has nothing to do with something thanklessly maintaining a library, and more to do with Artifactory situation and our own incident management.

Reviewer 4: I should've caught this in an earlier draft, but it stands out more as I see the rendering.

Reviewer: My thinking in originally considering the comic relevant was that "incident management" is actually often part of "maintaining a library"; they're both under the same umbrella of being responsible for sbt and ensuring its ongoing availability and viability.

Regardless, I definitely don't object to omitting it.

Another reason to omit it is that this blog post is really just an incident report and I think it should be kept short. The end of the post promises "We will let you know as soon as we settle on a plan". In _that_ blog post, perhaps the comic will be more appropriate, I don't know, we'll see.

Author: OK, I removed it in [0cc8d29](https://github.com/scala/scala-lang/pull/1492/commits/0cc8d29ba87e47e23c3ba572f952f986d5bc0ae8).

In practice, what do we rely on the sbt community repository for?

The sbt community repository is primarily used for two purposes:

1. It hosts “old” versions of sbt plugins that may still be used today in Scala projects.
   It is important to note that those sbt plugins are not necessarily that old. Some of
   them are stable and they simply didn't get a new release in the past two years,

>    meaning that they had no chances to migrate to the Central Maven repository.

Reviewer 3: ```suggestion
   meaning that they had no chance to migrate to the Central Maven repository.
```

2. It also hosts the Linux packages (deb and rpm) of sbt releases. Note that that part

>    is not read-only: we still publish Linux packages of new sbt releases there. These

Reviewer: ```suggestion
   is not read-only: we still publish new sbt releases there. These
```

Reviewer: But also, do you want to say "we"? It's Eugene that publishes them, not the Scala Center.

There might be other "we"s in the post?

   packages are used by all the Scala developers who install `sbt` via a Linux package
   manager (and by many Docker images that provide `sbt`).

> Last, it also hosts old releases of sbt (before the 1.x era) and Play framework.

Reviewer 3: ```suggestion
Lastly, it also hosts old releases of sbt (before the 1.x era) and Play framework.
```

For the record, here is the exhaustive list of URLs and their respective content:

- `https://repo.scala-sbt.org/scalasbt/sbt-plugin-releases` - sbt plugins for sbt 1.x, 0.13.x, 0.12.x etc.
- `https://repo.typesafe.com/typesafe/ivy-releases` - sbt main artifacts prior to sbt 1.x, and other artifacts such as Play.
- `https://repo.scala-sbt.org/scalasbt/debian` - sbt DEB installers.
- `https://repo.scala-sbt.org/scalasbt/rpm` - sbt RPM installers.

## Next Steps

> The outage lasted less than 24 hours and yet it impacted many developers. In

Reviewer 3: ```suggestion
The outage lasted less than 24 hours and impacted many developers. In
```
